Improve error about extra parens with parameter packs

With this code:

```swift
class Store<State> {
    var currentState: State { fatalError() }
}

class CompositeStore<CompositeState, each State, OriginalState>: Store<CompositeState> {
    let store: (repeat Store<each State>)
    let combine: ((repeat (each State)) -> CompositeState)

    override var currentState: CompositeState {
        return self.combine((repeat (each store).currentState))
    }

    init(store1: (repeat Store<each State>),
         combine: @escaping (repeat (each State)) -> CompositeState)
    {
        self.store = store1
        self.combine = combine
   }
}
```

The call to `self.combine` errors with:

```
/tmp/foo.swift:10:29: error: cannot pass value pack expansion to non-pack parameter of type 'repeat (each State)'
```

The core issue here is that the extra set of parentheses need to be removed. For other types this wouldn't be an issue so I think it would be nice if somehow that was potentially recommended in the diagnostic.

**Steps to reproduce**

```
xcrun --toolchain org.swift.59202306051a swiftc /tmp/foo.swift -Xfrontend -disable-availability-checking
```

**Environment**
- 6/5/23 main snapshot

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -7756,7 +7756,23 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     if (type1->isPlaceholder() || type2->isPlaceholder())
       return getTypeMatchSuccess();
 
+    // If parameter pack expansion contains more than one element and the other
+    // side is a tuple, record a fix.
     auto *loc = getConstraintLocator(locator);
+    if (loc->isLastElement<LocatorPathElt::ApplyArgToParam>()) {
+      if (auto packExpansion = type2->getAs<PackExpansionType>()) {
+        auto countType = simplifyType(packExpansion->getCountType(), flags);
+        if (auto paramPack = countType->getAs<PackType>()) {
+          if (type1->is<TupleType>() && paramPack->getNumElements() >= 1) {
+            if (recordFix(DestructureTupleToMatchPackExpansionParameter::create(
+                    *this, paramPack, loc))) {
+              return getTypeMatchFailure(loc);
+            }
+            return getTypeMatchSuccess();
+          }
+        }
+      }
+    }
     if (recordFix(AllowInvalidPackExpansion::create(*this, loc)))
       return getTypeMatchFailure(locator);
 
@@ -13473,14 +13489,16 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifySameShapeConstraint(
         auto paramPack = type2->castTo<PackType>();
 
         // Tailed diagnostic to explode tuples.
+        // FIXME: This is very similar to
+        // 'cannot_convert_single_tuple_into_multiple_arguments'; can we emit
+        // both of these in the same place?
         if (argPack->getNumElements() == 1) {
-          if (auto *tuple = argPack->getElementType(0)->getAs<TupleType>()) {
-            if (tuple->getNumElements() == paramPack->getNumElements()) {
-              return recordShapeFix(
-                  DestructureTupleToMatchPackExpansionParameter::create(
-                      *this, paramPack, loc),
-                  /*impact=*/2 * paramPack->getNumElements());
-            }
+          if (argPack->getElementType(0)->is<TupleType>() &&
+              paramPack->getNumElements() >= 1) {
+            return recordShapeFix(
+                DestructureTupleToMatchPackExpansionParameter::create(
+                    *this, paramPack, loc),
+                /*impact=*/2 * paramPack->getNumElements());
           }
         }
 

--- a/test/Constraints/pack-expansion-expressions.swift
+++ b/test/Constraints/pack-expansion-expressions.swift
@@ -617,3 +617,47 @@ do {
     _ = { (repeat test(x: each x)) }()
   }
 }
+
+// https://github.com/apple/swift/issues/66393
+do {
+  struct S<each T> {
+    var property: (repeat each T) -> Void { // expected-note 4 {{'property' declared here}}
+      get {}
+    }
+
+    func method(_: repeat each T) {} // expected-note 4 {{'method' declared here}}
+  }
+  S<Int, Bool>().method((5, true))
+  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+
+  S<Int, Bool>().method((5, true, 6))
+  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+
+  S<Int, Bool>().property((5, true))
+  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+
+  S<Int, Bool>().property((5, true, 6))
+  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+
+  func foo<each U>(u: repeat each U) {
+    S<repeat each U>().property((3, 4, 5))
+    // expected-error@-1 {{value pack expansion at parameter #0 expects 1 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+
+    // FIXME: The count of 'repeat each U' is not statically known, but error suggests that it is 1.
+    S<repeat each U>().method((3, 4, 5))
+    // expected-error@-1 {{value pack expansion at parameter #0 expects 1 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+    // FIXME: Bad diagnostics
+    // expected-error@-3 {{pack expansion requires that 'each U' and '_' have the same shape}}
+    // expected-error@-4 {{pack expansion requires that 'each U' and '_.RawValue' have the same shape}}
+
+    // FIXME: The count of '(Int, Int), repeat each U' is not statically known, but error suggests that it is 2.
+    S<(Int, Int), repeat each U>().method((3, 4))
+    // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+    // FIXME: Duplicate diagnostics
+    // expected-error@-3 2 {{pack expansion requires that 'each U' and '' have the same shape}}
+
+    // FIXME: The count of '(Int, Int), repeat each U' is not statically known, but error suggests that it is 2.
+    S<(Int, Int), repeat each U>().property((3, 4))
+    // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+  }
+}

